glean/src/core/upload/worker.ts

```diff
@@ -0,0 +1,179 @@
+
+// Error to be thrown in case the final ping body is larger than MAX_PING_BODY_SIZE.
+class PingBodyOverflowError extends Error {
+  constructor(message?: string) {
+    super(message);
+    this.name = "PingBodyOverflow";
+  }
+}
+
+class PingUploadWorker {
+  private currentJob?: Promise<void>;
+
+  constructor (
+    private readonly uploader: Uploader,
+    private readonly serverEndpoint: string,
+    private readonly policy = new Policy()
+  ) {}
+
+  /**
+   * Builds a ping request.
+   *
+   * This includes:
+   *
+   * 1. Includes Glean required headers to the ping;
+   *    These are the headers described in https://mozilla.github.io/glean/book/user/pings/index.html?highlight=headers#submitted-headers
+   * 2. Stringifies the body.
+   *
+   * @param ping The ping to include the headers in.
+   * @returns The updated ping.
+   */
+  private async buildPingRequest(ping: QueuedPing): Promise<{
+    headers: Record<string, string>,
+    payload: string | Uint8Array
+  }> {
+    let headers = ping.headers || {};
+    headers = {
+      ...ping.headers,
+      "Content-Type": "application/json; charset=utf-8",
+      "Date": (new Date()).toISOString(),
+      "X-Client-Type": "Glean.js",
+      "X-Client-Version": GLEAN_VERSION,
+      "X-Telemetry-Agent": `Glean/${GLEAN_VERSION} (JS on ${await Context.platform.info.os()})`
+    };
+
+    const stringifiedBody = JSON.stringify(ping.payload);
+    // We prefer using `strToU8` instead of TextEncoder directly,
+    // because it will polyfill TextEncoder if it's not present in the environment.
+    // Environments that don't provide TextEncoder are IE and most importantly QML.
+    const encodedBody = strToU8(stringifiedBody);
+
+    let finalBody: string | Uint8Array;
+    let bodySizeInBytes: number;
+    try {
+      finalBody = gzipSync(encodedBody);
+      bodySizeInBytes = finalBody.length;
+      headers["Content-Encoding"] = "gzip";
+    } catch {
+      finalBody = stringifiedBody;
+      bodySizeInBytes = encodedBody.length;
+    }
+
+    if (bodySizeInBytes > this.policy.maxPingBodySize) {
+      throw new PingBodyOverflowError(
+        `Body for ping ${ping.identifier} exceeds ${this.policy.maxPingBodySize}bytes. Discarding.`
+      );
+    }
+
+    headers["Content-Length"] = bodySizeInBytes.toString();
+    return {
+      headers,
+      payload: finalBody
+    };
+  }
+
+  /**
+   * Attempts to upload a ping.
+   *
+   * @param ping The ping object containing headers and payload.
+   * @returns The status number of the response or `undefined` if unable to attempt upload.
+   */
+  private async attemptPingUpload(ping: QueuedPing): Promise<UploadResult> {
+    try {
+      const finalPing = await this.buildPingRequest(ping);
+      return await this.uploader.post(
+        `${this.serverEndpoint}${ping.path}`,
+        finalPing.payload,
+        finalPing.headers
+      );
+    } catch(e) {
+      log(LOG_TAG, [ "Error trying to build ping request:", e ], LoggingLevel.Warn);
```

Comment:
Can the upload throw an error? If so I'd say:
```suggestion
      log(LOG_TAG, [ "Error trying to build or post ping request:", e ], LoggingLevel.Warn);
```